[FIX 36]: Manual Records

LeonieAdis opened the pull request:
This PR fixes #36. 

Now manual work records can be added to a budget.
A manual record has a description, an amount of money and a creation date. 

On the BudgetDetailsPage is a new link: Manual Records
Here you can see a table with all the manual records and a link to add a new manual record.

After a manual record was added, it is included in every chart, table, overview, etc that shows data of this budget (table in Budget Details, Burned Budget chart in Dashboard, table and chart in Week and Month Reports of one and multiple budgets, table in Budget Overview, table in Contract Overview, table and chart in Contract Details).

maximAtanasov commented on budgeteer-web-interface/src/main/java/org/wickedsource/budgeteer/persistence/manualRecord/ManualRecordEntity.java:
```diff
@@ -0,0 +1,89 @@
+    private String description;
+
+    @Column(nullable = false)
+    private int cents;
```
You used a `Money` object in the ManualRecord class, but an `int` here, why?

maximAtanasov commented on budgeteer-web-interface/src/main/java/org/wickedsource/budgeteer/persistence/manualRecord/ManualRecord.java:
```diff
@@ -0,0 +1,32 @@
+package org.wickedsource.budgeteer.persistence.manualRecord;
```
This class should probably be in the sevice layer.

maximAtanasov commented on budgeteer-web-interface/src/main/java/org/wickedsource/budgeteer/persistence/manualRecord/ManualRecordEntity.java:
```diff
@@ -0,0 +1,89 @@
+
+    @Temporal(TemporalType.DATE)
+    @Column(name="RECORD_DATE", nullable = false)
+    private Date date;
```
Records have dates (the date they were added). Maybe they should also have another date, like the date said action actually happend (maybe it could be optional). Then the user could see two columns in the table - Record Date and Date Added.

maximAtanasov commented on budgeteer-web-interface/src/main/java/org/wickedsource/budgeteer/persistence/record/WeeklyAggregatedRecordWithTaxBean.java:
```diff
@@ -3,6 +3,7 @@
 import lombok.Getter;
 import lombok.Setter;
 import org.joda.money.Money;
+import sun.nio.cs.ext.TIS_620;
```
What is this :D ?

maximAtanasov commented on budgeteer-web-interface/src/main/java/org/wickedsource/budgeteer/web/pages/budgets/manualRecords/overview/ManualRecordOverviewPage.java:
```diff
@@ -0,0 +1,58 @@
+package org.wickedsource.budgeteer.web.pages.budgets.manualRecords.overview;
```
I see no way to delete or edit existing records.

LeonieAdis commented:
I added all suggestions:
- manual records have now a "Charge Date", the date this action actually happend. By default, it's the creation date.
- manual records can be edited and deleted
- the money amount of a manual record is saved as a `Money`-object, not as an `int`

Additional, the calculation for the contract pages was not entirely correct, so I fixed that, too.

maximAtanasov commented:
Looks good :). I will merge this into master when the admin menu is merged first so that I don't have to modify the existing migration script in that branch.

maximAtanasov commented:
I will merge this into master and just won't update the admin-menu PR for now.

DK-Stern commented on budgeteer-web-interface/src/main/java/org/wickedsource/budgeteer/persistence/contract/ContractRepository.java:
```diff
@@ -20,56 +20,93 @@
 
-    @Query("select coalesce(sum(wr.minutes * wr.dailyRate/ 60 / 8),0) from WorkRecordEntity wr where wr.budget.contract.id = :contractId")
+    @Query("select (coalesce(sum(wr.minutes * wr.dailyRate/ 60 / 8),0) +" +
+           " (select coalesce(sum(record.moneyAmount),0) " +
```
This query doesn't work with oracle database!
So it's currently not possible to create new contracts, if you use a oracle database.

tinne replied:
This is what hibernate generates off the statement annotated on `getBudgetLeftByContractId` (pretty print by hand):

```sql
select nvl(sum(workrecord0_.minutes * workrecord0_.daily_rate / 60 / 8), 0) + ( 
    select nvl(sum(manualreco1_.money_amount), 0) 
    from manual_record_entity manualreco1_, budget budgetenti2_ 
    where manualreco1_.budget_id=budgetenti2_.id and 
        budgetenti2_.contract_id=?) as col_0_0_ 
from work_record workrecord0_ 
    cross join budget budgetenti3_ 
where workrecord0_.budget_id=budgetenti3_.id and 
    budgetenti3_.contract_id=?
```

The addition of seemingly independent information sets is thus integrated in a subselect clause where the inner one (select from manual_record and budget) may (but does not) depend on the outer query. In order for it to depend on the outer query in would however need to group the information by budget - note that the inner select happens on the cross product (any combination) of manual record and budget applicable to their join condition, and each thus combination would be applied to the outer select. This is simply not necessary, and in SQL this can be reformulated as such:

```sql
select nvl(sum(workrecord0_.minutes*workrecord0_.daily_rate/60/8), 0) + 
    nvl(sum(manualreco1_.money_amount), 0)
from work_record workrecord0_ 
cross join budget budgetenti3_ 
cross join manual_record_entity manualreco1_
where workrecord0_.budget_id=budgetenti3_.id and 
    manualreco1_.budget_id=budgetenti3_.id and
    budgetenti3_.contract_id=:contractId;
```

I'm currently not quite sure how this could be expressed in HQL so unless you wish to embed native SQL queries I suggest to split the statement in two more operable - I do not see any recognizable performance loss and it would hugely improve upon maintainability.

tinne replied:
Note that `Double getBudgetLeftByContractId(@Param("contractId") long contractId);` would not work as well due to the same problem.

LeonieAdis replied:
I couln't find a good  and working way to change the queries, so I split them up.
See my comment in #339 .
